libclamav/regex_list.c

```diff
@@ -660,43 +744,99 @@ static void list_add_tail(struct regex_list_ht *ht, struct regex_list *regex)
+    }
+    if (NULL == suffix) {
+        cli_errmsg("add_pattern_suffix: suffix must be initialized\n");
+        ret = CL_ENULLARG;
+        goto done;
+    }
+    if (NULL == iregex) {
+        cli_errmsg("add_pattern_suffix: iregex must be initialized\n");
+        ret = CL_ENULLARG;
+        goto done;
+    }
 
-    assert(matcher);
-    if (!regex) {
-        cli_errmsg("add_pattern_suffix: Unable to allocate memory for regex\n");
-        return CL_EMEM;
+    CLI_MALLOC(regex, sizeof(*regex),
+               cli_errmsg("add_pattern_suffix: Unable to allocate memory for regex\n");
+               ret = CL_EMEM);
+
+    if (NULL == iregex->pattern) {
+        regex->pattern = NULL;
+    } else {
+        CLI_STRDUP(iregex->pattern, regex->pattern,
+                   cli_errmsg("add_pattern_suffix: unable to strdup iregex->pattern");
+                   ret = CL_EMEM);
     }
-    regex->pattern = iregex->pattern ? cli_strdup(iregex->pattern) : NULL;
-    regex->preg    = iregex->preg;
-    regex->nxt     = NULL;
-    el             = cli_hashtab_find(&matcher->suffix_hash, suffix, suffix_len);
+    regex->preg = iregex->preg;
+    regex->nxt  = NULL;
+    el          = cli_hashtab_find(&matcher->suffix_hash, suffix, suffix_len);
     /* TODO: what if suffixes are prefixes of eachother and only one will
 	 * match? */
     if (el) {
         /* existing suffix */
-        assert((size_t)el->data < matcher->suffix_cnt);
+        if ((size_t)el->data >= matcher->suffix_cnt) {
+            cli_errmsg("add_pattern_suffix: el-> data too large");
+            ret = CL_ERROR;
+            goto done;
+        }
         list_add_tail(&matcher->suffix_regexes[(size_t)el->data], regex);
     } else {
         /* new suffix */
         size_t n    = matcher->suffix_cnt++;
         el          = cli_hashtab_insert(&matcher->suffix_hash, suffix, suffix_len, (cli_element_data)n);
         tmp_matcher = matcher->suffix_regexes; /*  save the current value before cli_realloc()	*/
         tmp_matcher = cli_realloc(matcher->suffix_regexes, (n + 1) * sizeof(*matcher->suffix_regexes));
+        /*TODO: switch to CLI_REALLOC when that is merged.*/
         if (!tmp_matcher) {
-            free(regex);
-            return CL_EMEM;
+            cli_errmsg("add_pattern_suffix: Unable to reallocate memory for matcher->suffix_regexes\n");
+            ret = CL_EMEM;
+            goto done;
         }
+
         matcher->suffix_regexes         = tmp_matcher; /*  success, point at new memory location   */
         matcher->suffix_regexes[n].tail = regex;
         matcher->suffix_regexes[n].head = regex;
-        if (suffix[0] == '/' && suffix[1] == '\0')
+        if (suffix[0] == '/' && suffix[1] == '\0') {
             matcher->root_regex_idx = n;
-        add_newsuffix(matcher, regex, suffix, suffix_len);
+        }
+
+        ret = add_newsuffix(matcher, regex, suffix, suffix_len);
+
+        /* Seeing CL_EMALFDB returned when calling this on daily.cvd,
```

Comment:
If there is a regex sig in daily that is bad and fails to load, we should drop it or fix it. 
The offending signature appears to be:
```
X:([^/?]+)?skrill\.com([/?])?:([^/?]+)?moneybookers\.com([/?])?
```
from daily.wdb. 

I manually tested with this `if()`-hack removed and that signature removed and no other signatures caused a failure. I'd rather drop that sig and delete this `if()`-hack. Let's talk to the malware team.

Reply:
For reference, the docs for `X:` WDB signatures: https://docs.clamav.net/manual/Signatures/PhishSigs.html?highlight=wdb#wdb-format

In my testing, we could resolve it by switching it to either:
`X:([^/?]+)?skrill\.com([/?])?:([^/?]+)?moneybookers\.com([/?].*)?`
or:
`X:([^/?]+)?skrill\.com([/?])?:([^/?]+)?moneybookers\.com`

I'll try to get it changed.